### PR TITLE
[fix] Allow setting sourceType in config file

### PR DESCRIPTION
**Pull request by alpreu**
### Motivation

Currently, there is an inconsistency between sinks and sources. Sinks can be created by passing the whole config in the sinkConfig file:

```
./bin/pulsar-admin sinks create --sink-config-file <file-name>.yaml
```

But Sources cannot. If we do the equivalent:

```
./bin/pulsar-admin sources create --source-config-file <file-name>.yaml
>Source archive not specified
```

To make it work, we currently have to explicitly pass in the type parameter:

```
./bin/pulsar-admin sources create --source-config-file <file-name>.yaml --source-type foo
```

### Modifications

Add the `sourceType` to `SourceConfig` so it can be passed through the config file too.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

- extended `TestCmdSources`

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [x] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/alpreu/pulsar/pull/9

### CODE REVIEW

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSources.java [hunk 1]
```diff
@@ -449,6 +449,8 @@ void processArguments() throws Exception {
 
             if (sourceType != null) {
                 sourceConfig.setArchive(validateSourceType(sourceType));
+            } else if (sourceConfig.getSourceType() != null) {
+                sourceConfig.setArchive(validateSourceType(sourceType));
```
Comment (nicoloboschi) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSources.java:453]:
```suggestion
                sourceConfig.setArchive(validateSourceType(sourceConfig.getSourceType()));
```

File: pulsar-client-tools/src/test/java/org/apache/pulsar/admin/cli/TestCmdSources.java [hunk 2]
```diff
@@ -412,6 +412,15 @@ public void testBatchSourceConfigMissingDiscoveryTriggererClassName() throws Exc
         expectedSourceConfig.setBatchSourceConfig(batchSourceConfig);
         testCmdSourceConfigFile(testSourceConfig, expectedSourceConfig);
     }
+
+    @Test(expectedExceptions = ParameterException.class, expectedExceptionsMessageRegExp = "Invalid source type 'foo' " +
```
Comment (nicoloboschi) [inline pulsar-client-tools/src/test/java/org/apache/pulsar/admin/cli/TestCmdSources.java:416]: could you add a happy path test ?

**Review by nicoloboschi — CHANGES_REQUESTED** [review]

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSources.java (hunk 1 above)
Comment (tisonkun) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSources.java]: Updated.

File: pulsar-client-tools/src/test/java/org/apache/pulsar/admin/cli/TestCmdSources.java (hunk 2 above)
Comment (tisonkun) [reply pulsar-client-tools/src/test/java/org/apache/pulsar/admin/cli/TestCmdSources.java]: I tried it:

```java

    @Test
    public void testCmdSourceConfigFileSourceType() throws Exception {
        ConnectorDefinition mockDefinition = new ConnectorDefinition();
        mockDefinition.setName("builtin");
        when(source.getBuiltInSources()).thenReturn(List.of(mockDefinition));

        SourceConfig sourceConfig = getSourceConfig();
        sourceConfig.setSourceType("builtin");

        SourceConfig exceptedConfig = getSourceConfig();
        exceptedConfig.setSourceType("builtin");
        exceptedConfig.setArchive("builtin://builtin");

        testCmdSourceConfigFile(sourceConfig, exceptedConfig);
    }
    
```

... while due to some dirty code inherit local run from cmd source the tests are hard to tune. That said, `validateSourceType` is inherited by `LocalSourceRunner` with total different manner.

**Review by tisonkun — APPROVED** [review]
LGTM.

**Review by nicoloboschi — APPROVED** [review]
LGTM

**Comment by tisonkun** [comment]
Merge the latest master and rerun...

**Comment by tisonkun** [comment]
Merging...

**Pull request merged**